File: Microsoft Office Developer Documentation Team/SharePoint 2010 101 Code Samples/49862-SharePoint 2010 101 Code Samples/SharePoint 2010 Creating Custom Timer Jobs/C#/GENERAL_CustomTimerJob/Features/CustomTimerJobFeature/CustomTimerJobFeature.EventReceiver.cs
using System;
using System.Runtime.InteropServices;
using System.Security.Permissions;
using Microsoft.SharePoint;
using Microsoft.SharePoint.Administration;
using Microsoft.SharePoint.Security;

namespace GENERAL_CustomTimerJob.Features.CustomTimerJobFeature
{
    /// <summary>
    /// For a custom timer job, you must use this feature receiver to install and configure
    /// the SPJobDefinition class
    /// </summary>
    /// <remarks>
    /// The GUID attached to this class may be used during packaging and should not be modified.
    /// </remarks>

    [Guid("1b45bf8b-f5e2-4155-8f7f-999b1163c3d1")]
    public class CustomTimerJobFeatureEventReceiver : SPFeatureReceiver
    {
        const string TIMER_JOB_NAME = "DemoTimerJob";

        public override void FeatureActivated(SPFeatureReceiverProperties properties)
        {
            //In this event we create and install the timer job
            //Start by finding the SPSite.
            SPSite site = (SPSite)properties.Feature.Parent;
            //Make sure the timer job isn't already registered
            foreach (SPJobDefinition job in site.WebApplication.JobDefinitions)
            {
                if (job.Name == TIMER_JOB_NAME)
                {
                    job.Delete();
                }
            }
            //Create a new Timer job
            GENERAL_CustomTimerJob newTimerJob = new GENERAL_CustomTimerJob(TIMER_JOB_NAME, site.WebApplication);
            //Configure the schedule and save it
            SPMinuteSchedule jobSchedule = new SPMinuteSchedule();
            jobSchedule.BeginSecond = 0;
            jobSchedule.EndSecond = 59;
            jobSchedule.Interval = 5;
            newTimerJob.Schedule = jobSchedule;
            newTimerJob.Update();
        }


        public override void FeatureDeactivating(SPFeatureReceiverProperties properties)
        {
            //In this event we must clean up by deleting the timer job
            SPSite site = properties.Feature.Parent as SPSite;
            //Locate the right timer job
            foreach (SPJobDefinition job in site.WebApplication.JobDefinitions)
            {
                if (job.Name == TIMER_JOB_NAME)
                {
                    //This one is the right job. Delete it.
                    job.Delete();
                }
            }

        }
    }
}

File: Microsoft Office Developer Documentation Team/SharePoint 2010 101 Code Samples/[VB]-SharePoint 2010 101 Code Samples/SharePoint 2010 Creating Custom Timer Jobs/VB/GENERAL_CustomTimerJob/Features/CustomTimerJobFeature/CustomTimerJobFeature.EventReceiver.vb
Option Explicit On
Option Strict On

Imports System
Imports System.Runtime.InteropServices
Imports System.Security.Permissions
Imports Microsoft.SharePoint
Imports Microsoft.SharePoint.Administration
Imports Microsoft.SharePoint.Security

''' <summary>
''' For a custom timer job, you must use this feature receiver to install and configure
''' the SPJobDefinition class
''' </summary>
''' <remarks>
''' The GUID attached to this class may be used during packaging and should not be modified.
''' </remarks>

<GuidAttribute("d7faf5a5-9d5e-41c3-b768-8a30c6136990")> _
Public Class CustomTimerJobFeatureEventReceiver 
    Inherits SPFeatureReceiver

    Const TIMER_JOB_NAME As String = "DemoTimerJob"

    Public Overrides Sub FeatureActivated(ByVal properties As SPFeatureReceiverProperties)
        'In this event we create and install the timer job
        'Start by finding the SPSite.
        Dim site As SPSite = DirectCast(properties.Feature.Parent, SPSite)
        'Make sure the timer job isn't already registered
        For Each job As SPJobDefinition In site.WebApplication.JobDefinitions
            If job.Name = TIMER_JOB_NAME Then
                job.Delete()
            End If
        Next
        'Create a new Timer job
        Dim newTimerJob As GENERAL_CustomTimerJob = New GENERAL_CustomTimerJob(TIMER_JOB_NAME, site.WebApplication)
        'Configure the schedule and save it
        Dim jobSchedule As SPMinuteSchedule = New SPMinuteSchedule()
        jobSchedule.BeginSecond = 0
        jobSchedule.EndSecond = 59
        jobSchedule.Interval = 5
        newTimerJob.Schedule = jobSchedule
        newTimerJob.Update()
    End Sub

    Public Overrides Sub FeatureDeactivating(ByVal properties As SPFeatureReceiverProperties)
        'In this event we must clean up by deleting the timer job
        Dim site As SPSite = DirectCast(properties.Feature.Parent, SPSite)
        'Locate the right timer job
        For Each job As SPJobDefinition In site.WebApplication.JobDefinitions
            If job.Name = TIMER_JOB_NAME Then
                'This one is the right job. Delete it.
                job.Delete()
            End If
        Next
    End Sub

End Class
